src/android/usb_host/usb_device.cpp

```diff
@@ -0,0 +1,276 @@
+                    if (h->bDescriptorType == USB_DT_INTERFACE_ASSOCIATION) {
+                        usb_interface_association a(i, (usb_interface_assoc_descriptor *) h);
+                        configuration.add_interface_association(a);
+                    } else if (h->bDescriptorType == (USB_DT_INTERFACE & ~USB_TYPE_MASK)) {
+                        i++;
+                        usb_interface_descriptor *interface_descriptor = (usb_interface_descriptor *) h;
+                        usb_interface usbInterface(*interface_descriptor);
+                        for (int e = 0; e < interface_descriptor->bNumEndpoints;) {
+                            h = usb_descriptor_iter_next(&it);
+                            if (h->bDescriptorType == (USB_DT_ENDPOINT & ~USB_TYPE_MASK)) {
+                                e++;
+                                usb_endpoint usb_endpoint(*(usb_endpoint_descriptor *) h);
+                                usbInterface.add_endpoint(usb_endpoint);
+                                auto pipe = std::make_shared<usb_pipe>(_handle, usb_endpoint);
+                                add_pipe(usb_endpoint.get_endpoint_address(), pipe);
+                            }
+                        }
+                        configuration.add_interface(usbInterface);
+                    }
+                    h = usb_descriptor_iter_next(&it);
+                }
+
+                add_configuration(configuration);
+
+            } while (h != nullptr && (h->bDescriptorType == USB_DT_CONFIG));
+        } else {
+            return;
+        }
+    }
+}
+
+const std::string& device::get_device_description() const {
+    auto desc = _usb_device_descriptor;
+    std::stringstream ss;
+
+    ss << "Type: " << (int) desc->bDescriptorType << " ("
+       << descriptorTypeToString(desc->bDescriptorType) << ")\n"
+       << "   Length: " << (int) desc->bLength << std::endl
+       << "   USB version: " << std::hex << (int) __le16_to_cpu(desc->bcdUSB) << std::dec
+       << std::endl
+       << "   Device version: " << (int) __le16_to_cpu(desc->bcdDevice) << std::endl
+       << "   Class: " << (int) desc->bDeviceClass << " ("
+       << classToString(desc->bDeviceClass) << ")\n"
+       << "   Subclass: " << (int) desc->bDeviceSubClass << std::endl
+       << "   Protocol: " << std::hex << (int) desc->bDeviceProtocol << std::dec
+       << std::endl
+       << "   Num configs: " << (int) desc->bNumConfigurations << std::endl
+       << "   Product ID: " << (int) __le16_to_cpu(desc->idProduct) << std::endl
+       << "   Vendor ID: " << (int) __le16_to_cpu(desc->idVendor) << std::endl
+       << "   Product: " << (int) desc->iProduct << " ("
+       << get_product() << ")\n"
+       << "   Manufacturer: " << (int) desc->iManufacturer << " ("
+       << get_manufacturer() << ")\n"
+       << "   Serial num: " << (int) desc->iSerialNumber << " ("
+       << get_serial_number() << ")\n\n";
+    return ss.str();
```

Comment:
When building, this line produces a warning because the function is returning a reference to a local temporary. This function ought to return a string rather than reference to one.

Reply:
Fixed in 8138bfb2f